### PR TITLE
Added dropdown to display ads / billboards

### PR DESCRIPTION
**Pull request by rt4914**
## What type of PR is this? (check all applicable)

- [x] Feature

## Description

This PR adds the org name and hardcoded dropdown menu to all currently existing billboards.

## Related Tickets & Documents

- Closes https://github.com/forem/forem-internal-eng/issues/531

## QA Instructions, Screenshots, Recordings

**Reference Video**: https://www.loom.com/share/92859e41b05745b99423a653d83cc5f9

**Figma Designs**: https://www.figma.com/file/9UfSjNYJs745w3Q8JCUqsL/Small-Wins---%E2%80%A8Core-Experience-%E2%9A%A1%EF%B8%8F?node-id=1157%3A5067&t=d2xqTG6IUYL6oZq9-0

**How to test?**
1. Sign in as admin locally.
2. Create a `billboard` page on http://localhost:3000/admin/customization/pages
3. Make sure that the value of `slug` is `billboards` and select `Is top level path?`
4. Once finished, visit http://localhost:3000/billboard to see that the page is visible, if not, there was some mistake in creating the page.
5. Now visit http://localhost:3000/admin/customization/display_ads to create different types of ads (5 types of ads a possible) and check them on either home page or on article page.
6. Once you visit any billboard/display ad, click on three-dots/overflow-menu icon to see the dropdown menu.
7. Click on the two available options of dropdown menu to see if they work correctly. 

### Output
Screenshot 2023-02-24 at 2 23 36 PM Screenshot 2023-02-24 at 2 24 05 PM
Screenshot 2023-02-24 at 2 23 41 PM Screenshot 2023-02-24 at 2 24 11 PM

Latest UI Output: https://www.loom.com/share/7c80770fd75b4be79e531c8731d52f7e

Screenshot 2023-02-24 at 2 08 04 PM
Screenshot 2023-02-24 at 2 25 40 PM
Screenshot 2023-02-24 at 2 08 29 PM
Screenshot 2023-02-24 at 2 08 37 PM
Screenshot 2023-02-24 at 2 08 46 PM

### UI accessibility concerns?

Improves accessibility by making users understand what a billboard is and access to settings.

## Added/updated tests?

- [x] No tests added https://github.com/forem/forem/pull/19106#pullrequestreview-1301412835

## [Forem core team only] How will this change be communicated?

_Will this PR introduce a change that impacts Forem members or creators, the
development process, or any of our internal teams? If so, please note how you
will share this change with the people who need to know about it._

- [x] I will share this change internally with the appropriate teams

## [optional] Are there any post deployment tasks we need to perform?

@maestromac Can you deploy this on canary1 or canary2 for testing?

### CODE REVIEW

**Comment by rt4914** [comment]
Hi @jaw6
I need help in two things:
1. Can you check the code once and suggest anything which needs to be changed or if the code if going in right direction or not? (Maybe focus on naming of files, classes, ids as it was a bit confusing between display-ad, sponsorship and billboards).
2. Also I need help with writing test cases -- unable to identify what would be the best file to write test case for this --- also any references from codebase which can help in writing test cases for this functionality?

File: config/locales/fr.yml
Comment (rt4914) [inline config/locales/fr.yml:153]: Not sure about these translations.

File: app/views/articles/_sidebar.html.erb
Comment (rt4914) [inline app/views/articles/_sidebar.html.erb:13]: Showing this name too but this is not part of this issue, so might remove it later.

File: app/views/shared/_display_ad_dropdown_content.html.erb
Comment (jaw6) [inline app/views/shared/_display_ad_dropdown_content.html.erb:2]: This is doing exactly what the issue requests, and yet, with this link as a 404, we probably shouldn't deploy this. What about adding something so that it becomes valid as soon as `/billboard` exists? Something like:

```suggestion
        <% if Page.exists?(slug: "billboard") %>
```

We'd have a dropdown with only one item, but I think that'd be better than a link that goes nowhere. What do you think?

File: app/views/shared/_display_ad_dropdown_content.html.erb
Comment (jaw6) [inline app/views/shared/_display_ad_dropdown_content.html.erb:9]: One possible enhancement here (or later) might be if we had `#anchors` on the settings page, we could link directly to the sponsorship settings (since, otherwise, the user might not realize where on that page the settings could be found, and text search isn't helpful).

**Review by jaw6 — APPROVED** [review]
@rt4914 this looks great! No concerns from me on naming classes, files, or ids. Re-naming everything `display_ads -> billboards` will be a major effort, I can understand how it would be confusing, but I think this is all fine. I made two minor suggestions that might help, but take them or leave them as you like.

As for testing: I don't see a lot of value in testing this on the backend, I would think that we'd test `setupDisplayAdDropdown` in Cypress. Even then, the value of that testing is pretty minimal. (For what it's worth, my usual test-value-gut-check is "if this were broken, how quickly would we mobilize to fix it?" If this dropdown didn't appear when the button were clicked, I would guess we would treat it as a very minor bug.)

**Comment by mirie** [comment]
@lboogie04 can you review this PR too?

**Comment by mirie** [comment]
@lboogie04 can you review this PR too?

File: app/views/articles/_sidebar.html.erb
Comment (benhalpern) [inline app/views/articles/_sidebar.html.erb:11]: It looks like more or less the same code in five places here, I think this needs to be pulled into a partial.

File: app/views/shared/_display_ad_dropdown_content.html.erb
Comment (rt4914) [reply app/views/shared/_display_ad_dropdown_content.html.erb]: Done. I have added this check and also verified that it does work.
Screenshot 2023-02-20 at 2 39 44 PM

File: app/views/shared/_display_ad_dropdown_content.html.erb
Comment (rt4914) [reply app/views/shared/_display_ad_dropdown_content.html.erb]: Done. I have added `id=sponsors" to the destination content and updated this URL accordingly.

File: app/views/articles/_sidebar.html.erb
Comment (rt4914) [reply app/views/articles/_sidebar.html.erb:11]: @benhalpern I was unable to remove this entire display-ad code to a partial because not all display ads as same in design, especially the below-comments-ad is much different and therefore what I have done is I have taken the entire `header` code of the display ad into a partial. Earlier we had only the contents of drop-down menu into the partial.

**Comment by pkfrank** [comment]
Let's not merge this until we:

- [ ] sort out and fix the final design requirements
- [ ] update the Community Name on DEV

File: app/views/articles/_sidebar.html.erb
Comment (rt4914) [reply app/views/articles/_sidebar.html.erb]: Update on this @benhalpern I have merged the display ad fully.

File: app/models/display_ad_event.rb
Comment (rt4914) [inline app/models/display_ad_event.rb:14]: @jaw6 I need your suggestion here.

I have added `CONTEXT_TYPE_ARTICLE` which gets used for display ads which are shown on articles. Earlier we were not using any context for those display ads. With this implementation we have 3-display-ads on home page with 'home' context and two on article page with `article` context.
So first, is this correct?
Why are we not using 5 different context to track the impressions in more depth, like `home-left-1`, `home-left-2`, `home-right`, `article-right` and `article-below-comment`

**Comment by rt4914** [comment]
@jaw6 I have updated the code slightly and I need you attention on this comment https://github.com/forem/forem/pull/19106/files#r1114326574

**Comment by rt4914** [comment]
> * sort out and fix the final design requirements

@pkfrank I have finished ` sort out and fix the final design requirements` task from above comment and it can be verified here too https://www.loom.com/share/7c80770fd75b4be79e531c8731d52f7e

**Comment by rt4914** [comment]
@anujbhavsar96 Can you check the video and verify if its good to be deployed on canary again https://www.loom.com/share/7c80770fd75b4be79e531c8731d52f7e

**Comment by anujbhavsar96** [comment]
The dropdown looks perfect now @rt4914 

Just one more thing, as you can see below, there are some spacing issues _(Above and below content)_. We can have consistent styling and spacing across all the billboard units, so they look the same as a component.

image

**Comment by rt4914** [comment]
> The dropdown looks perfect now @rt4914
> 
> Just one more thing, as you can see below, there are some spacing issues _(Above and below content)_. We can have consistent styling and spacing across all the billboard units, so they look the same as a component.
> 
> image

@anujbhavsar96 Created this quick video explaining why this issue is happening, how can we solve it and actually if we can solve it or not. https://loom.com/share/74e4bd06925f4bd19b20065c89cdf6e3

File: app/models/display_ad_event.rb
Comment (jaw6) [reply app/models/display_ad_event.rb:14]: It's a good question! I don't really know _why_, or rather _how_ or even _if_ `context` is used. That attribute has been frozen as `"home"` since [the very beginning](https://github.com/forem/forem/pull/3589/files#diff-a2624b7194585f380b666f690e7c304c61212c39a1f30b397e1c5d1650e27886R214) -- [later PRs](https://github.com/forem/forem/pull/15946/) tweaked the mechanics but we did not change `DisplayAdEvent` even [when we introduced new placement areas](https://github.com/forem/forem/pull/18141), as it wasn't part of [the ask](https://github.com/forem/forem-internal-eng/issues/486) (and I was too :new: here to know `DisplayAdEvent` even existed).

I _think_ a reason that it hasn't been changed is that I don't think it gives us much information beyond what we have _already_ in `DisplayAd#placement_area` -- or, phrased somewhat differently, I think `context` was something @benhalpern assumed "would be" valuable [at the beginning](https://github.com/forem/forem/pull/3589/files#diff-a2624b7194585f380b666f690e7c304c61212c39a1f30b397e1c5d1650e27886R214) and maybe it _isn't_ valuable, but also maybe it isn't dragging us down, so removing it hasn't been a priority, but neither has "enhancing it", so to speak.

But most of that happened well before my time here, I am guessing at a lot here!

I'll defer to @benhalpern and @mirie, my sense is that it probably isn't worth changing in _this_ PR, but we should maybe open a new issue to discuss / prioritize it separately.

**Comment by rt4914** [comment]
> The dropdown looks perfect now @rt4914
> 
> Just one more thing, as you can see below, there are some spacing issues _(Above and below content)_. We can have consistent styling and spacing across all the billboard units, so they look the same as a component.
> 
> image

Update: This has been solved after discussing with @anujbhavsar96. We have removed `margin-top` for h1-h6, a and img tags. For this I have created new css class `text-styles--display-ad` which overrides these tags to have margin-top=0.

**Review by jaw6 — APPROVED** [review]
Changes look good to me -- I think we'll want to figure out a long-term plan for `context_type`, but I don't think that needs to hold up this work :+1:

**Review by anujbhavsar96 — APPROVED** [review]
Design changes look good to me🚀

**Review by benhalpern — APPROVED** [review]
This looks good to me. This ultimately did create an adjustment in how the overall padding in the below comments area looks, and we may need to continue to futz with this to get it just right, but I think it's in a good place to move forward.

**Pull request merged**
